File: frontend/src/types/role.ts
export interface ApiPermission {
  id: string
  name: string
  description: string
  created_at: string
  updated_at: string
}

export interface ApiRoleMenuPermission {
  menu_id: string
  menu_name: string
  permissions: ApiPermission[]
}

export interface ApiRole {
  id: string
  name: string
  description: string
  is_active: boolean
  created_at: string
  updated_at: string
  menu_permissions?: ApiRoleMenuPermission[]
}

export interface CreateRolePayload {
  name: string
  description: string
  is_active: boolean
}

export interface UpdateRolePayload {
  name?: string
  description?: string
  is_active?: boolean
}

export interface AssignPermissionsItem {
  menu_id: string
  permission_ids: string[]
}

export interface AssignPermissionsPayload {
  items: AssignPermissionsItem[]
}

File: backend/internal/features/role/domain/role.go
package domain

import (
	"context"
	"time"

	permissiondomain "apigofiberhorpug/internal/features/permission/domain"
)

type Role struct {
	ID              string               `json:"id"`
	Name            string               `json:"name"`
	Description     string               `json:"description"`
	IsActive        bool                 `json:"is_active"`
	CreatedAt       time.Time            `json:"created_at"`
	UpdatedAt       time.Time            `json:"updated_at"`
	MenuPermissions []RoleMenuPermission `json:"menu_permissions,omitempty"`
}

type RoleMenuPermission struct {
	MenuID      string                        `json:"menu_id"`
	MenuName    string                        `json:"menu_name"`
	Permissions []permissiondomain.Permission `json:"permissions"`
}

type CreateRoleRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	IsActive    *bool  `json:"is_active"`
}

type UpdateRoleRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	IsActive    *bool  `json:"is_active"`
}

type RoleMenuPermissionItem struct {
	MenuID        string   `json:"menu_id"`
	PermissionIDs []string `json:"permission_ids"`
}

type AssignPermissionsRequest struct {
	Items []RoleMenuPermissionItem `json:"items"`
}

type RoleRepository interface {
	FindByID(ctx context.Context, id string) (*Role, error)
	List(ctx context.Context, limit, offset int) ([]*Role, error)
	ListActive(ctx context.Context) ([]*Role, error)
	Count(ctx context.Context) (int, error)
	Create(ctx context.Context, role *Role) error
	Update(ctx context.Context, role *Role) error
	Delete(ctx context.Context, id string) error
	AssignMenuPermissions(ctx context.Context, roleID string, items []RoleMenuPermissionItem) error
	GetMenuPermissions(ctx context.Context, roleID string) ([]RoleMenuPermission, error)
}
